src/librustc_data_structures/obligation_forest/mod.rs

```diff
@@ -129,57 +163,64 @@ impl<O: Debug, T: Debug> ObligationForest<O, T> {
     }
 
     pub fn start_snapshot(&mut self) -> Snapshot {
-        self.snapshots.push(self.trees.len());
+        self.snapshots.push(SnapshotData {
+            node_len: self.nodes.len(),
+            cache_list_len: self.cache_list.len()
+        });
         Snapshot { len: self.snapshots.len() }
     }
 
     pub fn commit_snapshot(&mut self, snapshot: Snapshot) {
         assert_eq!(snapshot.len, self.snapshots.len());
-        let trees_len = self.snapshots.pop().unwrap();
-        assert!(self.trees.len() >= trees_len);
+        let info = self.snapshots.pop().unwrap();
+        assert!(self.nodes.len() >= info.node_len);
+        assert!(self.cache_list.len() >= info.cache_list_len);
     }
 
     pub fn rollback_snapshot(&mut self, snapshot: Snapshot) {
         // Check that we are obeying stack discipline.
         assert_eq!(snapshot.len, self.snapshots.len());
-        let trees_len = self.snapshots.pop().unwrap();
+        let info = self.snapshots.pop().unwrap();
 
-        // If nothing happened in snapshot, done.
-        if self.trees.len() == trees_len {
-            return;
+        for entry in &self.cache_list[info.cache_list_len..] {
+            self.done_cache.remove(entry);
+            self.waiting_cache.remove(entry);
         }
 
-        // Find root of first tree; because nothing can happen in a
-        // snapshot but pushing trees, all nodes after that should be
-        // roots of other trees as well
-        let first_root_index = self.trees[trees_len].root.get();
-        debug_assert!(self.nodes[first_root_index..]
-                          .iter()
-                          .zip(first_root_index..)
-                          .all(|(root, root_index)| {
-                              self.trees[root.tree.get()].root.get() == root_index
-                          }));
-
-        // Pop off tree/root pairs pushed during snapshot.
-        self.trees.truncate(trees_len);
-        self.nodes.truncate(first_root_index);
+        self.nodes.truncate(info.node_len);
+        self.cache_list.truncate(info.cache_list_len);
     }
 
     pub fn in_snapshot(&self) -> bool {
         !self.snapshots.is_empty()
     }
 
-    /// Adds a new tree to the forest.
+    /// Registers an obligation
     ///
-    /// This CAN be done during a snapshot.
-    pub fn push_tree(&mut self, obligation: O, tree_state: T) {
-        let index = NodeIndex::new(self.nodes.len());
-        let tree = TreeIndex::new(self.trees.len());
-        self.trees.push(Tree {
-            root: index,
-            state: tree_state,
-        });
-        self.nodes.push(Node::new(tree, None, obligation));
+    /// This CAN be done in a snapshot
+    pub fn register_obligation(&mut self, obligation: O) {
+        self.register_obligation_at(obligation, None)
+    }
+
+    fn register_obligation_at(&mut self, obligation: O, parent: Option<NodeIndex>) {
+        if self.done_cache.contains(obligation.as_predicate()) { return }
```

Comment:
Parents can't be in a `Pending` state, but they can be in any other intermediate state.